eclipse-theia/theia
# Use ScmTreeWidget in both commit details and in the diff widget

fixes https://github.com/eclipse-theia/theia/issues/7616

#### What it does

This PR updates both the following to use ScmTreeWidget.
- Commit Details editor (opened from the History view)
- Diff view (opened from "Compare With..." on a folder's context menu)

This provides a more consistent view of resource changes, regardless of whether they are uncommitted, in a commit, or are a result of a diff between any two versions of a folder.  Note that these widgets are provided by Theia's Git package, so not available if using the VS Code Git Builtin.

Some additional functionality comes along with this:
- The diffs can be opened from the Commit Details

The Diff widget had two buttons, <- and ->, to step through changes within a file.  Those buttons no longer exist and one now uses the cursor left and cursor right keys to step through changes.  That brings it in line with the Source Control view and is also, with this PR, now the behavior in the Commit Details widget.  If anyone wants to keep those buttons then that is fine (let me know) but we should probably also added them to both the Source Control and the Commit Details view.

Note that NavigableListWidget is now used only by the History widget.  When the History is incorporated into the Timeline then NavigableListWidget will go away.

Note there is a known bug in this PR.  If one opens a Diff on a folder, then makes a change to the files or staging that would change the Diff, the Diff widget does not immediately update.  If one resizes the Diff widget then the refresh occurs.  The forceUpdate is being done in ScmTreeWidget so I suspect this is some weird interaction with ReactVirtualized.  I'm hoping someone will know what the cause may be, otherwise I will spend more time looking at it.

#### How to test

- Test both the Commit Details widget and the Diff view.
- Check that the Source Control view looks and behaves exactly as before.
- Check in all three places that the tree/list toggle button works correctly.

#### Review checklist

- [x ] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
